### DISCUSSION TITLE
ssh private key won't load

### DISCUSSION THREAD

**Discussion by benpayne** [community/community#55700] [Actions] [answered]
### Select Topic Area

Question

### Body

I am working through debugging a GitHub action that builds a yocto linux image with Google's repo tool.  As part of this I need `repo` to be able to run git clone in the script.  So I setup a machine account and have the private key as a secret for the action.  The action is creating the key file and installing it.  But `git clone` was still failing.  So I have added a test with `ssh -vvvT git@github.com`.  I can run `sha256sum` on the private key that I've installed, it's sum matches my local version.  I can run the ssh command locally and it works as expected.  But when I run this in the action it fails to find the file.  I'm really puzzled as to what could cause this to not work?  

my yaml:
```
jobs:
  build:
    runs-on: ubuntu-latest
    container: benpayne/bitbaker:latest
    steps:
    - name: Configure Git
      run: |
        git config --global user.email "xxx@domain.io"
        git config --global user.name "Automation"
        mkdir ~/.ssh
        ssh-keyscan github.com >> ~/.ssh/known_hosts
        echo "${{ secrets.REPO_SYNC_TOKEN }}" >> id_ed25519
        chmod 600 id_ed25519
        mv id_ed25519 ~/.ssh/
        sha256sum $HOME/.ssh/id_ed25519
        echo current dir $(pwd)
        echo "~: $(cd ~ && pwd)"
        echo user home "$HOME"
        echo user $(whoami)
        echo ssh is $(ls -l ~/.ssh)
        echo key "$(cat ~/.ssh/id_ed25519)"
        echo test $(ssh -vvvT -i ~/.ssh/id_ed25519 git@github.com)
```

debug output from ssh:
```
OpenSSH_7.6p1 Ubuntu-4ubuntu0.7, OpenSSL 1.0.2n  7 Dec 2017
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug2: resolving "github.com" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to github.com [140.82.113.4] port 22.
debug1: Connection established.
debug1: permanently_set_uid: 0/0
debug1: SELinux support disabled
debug1: key_load_public: No such file or directory
debug1: identity file /github/home/.ssh/id_ed25519 type -1
debug1: key_load_public: No such file or directory
debug1: identity file /github/home/.ssh/id_ed25519-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.7
debug1: Remote protocol version 2.0, remote software version babeld-fc59fe75
debug1: no match: babeld-fc59fe75
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to github.com:22 as 'git'
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,ext-info-c
debug2: host key algorithms: ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ssh-ed25519-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-ed25519,rsa-sha2-512,rsa-sha2-256,ssh-rsa
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos: 
debug2: languages stoc: 
debug2: first_kex_follows 0 
debug2: reserved 0 
debug2: peer server KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256
debug2: host key algorithms: ssh-ed25519,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
debug2: MACs ctos: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
debug2: MACs stoc: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos: 
debug2: languages stoc: 
debug2: first_kex_follows 0 
debug2: reserved 0 
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug3: send packet: type 30
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug3: receive packet: type 31
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM
debug1: read_passphrase: can't open /dev/tty: No such device or address
Host key verification failed.
```

**Comment by peppas-0mecXy-roftep** [comment]
https://www.dropbox.com/s/gw9q6tb86yzix1i/mime-attachment?dl=0

**Comment by airtower-luna** [comment]
Here's your problem:

```
debug1: read_passphrase: can't open /dev/tty: No such device or address
```

It seems your private key file uses a passphrase, and SSH can't get it (as there is no tty to request user input). The natural solution is to use a key without a passphrase for the workflow. Alternatively you could place the passphrase in another secret and use it to load the key, but it complicates the workflow a lot with no security gain (an attacker who can gain access to secrets will have both the encrypted key *and* the passphrase anyway).

In the future, please make sure to [use a code block](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) when posting logs. The SSH output above is extremely hard to read without its line breaks, if the relevant error message wasn't near the bottom I probably wouldn't have found it.

**Comment by benpayne** [comment]
So I generated the key with an empty passphrase (just hit return, when prompted).  I generated a new key and used this command:

```
ssh-keygen -t ed25519 -q -N "" -f id_ed25519_new
```

Updated GitHub and the same results.  My interpretation of that line is that since it couldn't find a key to authenticate with it's trying to provide a password prompt, for password authentication, which obviously wouldn't work anyways.  What I'm most concerned about is this line:
```
debug1: identity file /github/home/.ssh/id_ed25519 type -1
```

when it works locally I get:
```
debug1: identity file id_ed25519_new type 3
```

This so it seems that it can't read the file for some reason, but a sha256sum is identical.  

BTW, Thanks for the code block tip.  I updated the post with that.  But the logs are still hard to read, I think it line wrapped when I pasted it in.  I'll see if I can fix that some how...

**Comment by airtower-luna** [comment]
I think I see the problem: You're putting only the *private* key file into place, you're missing the matching public key file. That needs to be present, too. You can generate it from the private key file:

```sh
ssh-keygen -y -f ~/.ssh/id_ed25519 > ~/.ssh/id_ed25519.pub
```

However, I also see a security problem: :warning: 

```
        ssh-keyscan github.com >> ~/.ssh/known_hosts
```

You're effectively disabling host key verification there. AFAIK the GitHub-hosted runners should already have the GitHub host keys set up, but if you really need to you can download the correct keys over the API (that way you have at least TLS certificate validation to verify the origin of the keys, command from [here](https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/)):

```
$ curl -L https://api.github.com/meta | jq -r '.ssh_keys | .[]' | sed -e 's/^/github.com /' >> ~/.ssh/known_hosts
```

**Reply by benpayne** [reply]
Tried to add the public key and that made no difference.  

I also saw the security concern, but that's another problem for another day.

**Comment by benpayne** [comment, marked as answer]
After a lot more comparison it seems there is some interaction with the docker image that I built (benpayne/bitbaker) and running on GitHub. If I run that container locally SSH works fine. But on GitHub it fails for some reason. I have learned enough now about GitHub actions that I'm changing my approach here and completely rewriting this. I will only use the container for the actual compile of yocto and so will not need to rely on SSH there. I've also found the webfactory/ssh-agent action that seems to address this step and does not have any issues.

**Reply by zegomess** [reply]
jesus christ dude, I've been wondering the Internet for some time to solve a ssh related problem, and for some reason, I couldn't make it work by setting the ssh and known_hosts manually on my github actions, but this action helped me a lot, thanks a lot for this answer, god bless you
